catalog/glue.go

```diff
@@ -0,0 +1,162 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package catalog
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/apache/iceberg-go/io"
+	"github.com/apache/iceberg-go/table"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/service/glue"
+	"github.com/aws/aws-sdk-go-v2/service/glue/types"
+)
+
+var (
+	_ Catalog = (*GlueCatalog)(nil)
+)
+
+type GlueAPI interface {
+	GetTable(ctx context.Context, params *glue.GetTableInput, optFns ...func(*glue.Options)) (*glue.GetTableOutput, error)
+	GetTables(ctx context.Context, params *glue.GetTablesInput, optFns ...func(*glue.Options)) (*glue.GetTablesOutput, error)
+}
+
+type GlueCatalog struct {
+	glueSvc GlueAPI
+}
+
+func NewGlueCatalog(awscfg aws.Config) *GlueCatalog {
+	return &GlueCatalog{
+		glueSvc: glue.NewFromConfig(awscfg),
+	}
+}
```

Comment:
Rather than `NewGlueCatalog` like this, maybe we follow the more typical iceberg setup (similar to pyiceberg) and create a `LoadCatalog` function taking a uri + map[string]string (properties/options) and return the `Catalog` interface directly.

We can then infer which type of catalog to create based on the uri or from a `type` property. I have an example of this in my older iceberg implementation of stuff located here: https://github.com/zeroshade/icegopher/blob/main/table/catalog/catalog.go also containing an implementation for the rest catalog there.

Thoughts?

Reply:
@zeroshade yeah this abstraction seems a bit more than I would expect in a Go library, I prefer the more common functional options where you inject specific overrides. https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis 

Providing maps of properties just result in you needing to re-implement the myriad of overrides provided by say the aws.Config, or similar configs for other providers. 

Creating an abstraction where "everything is created" feels a bit unnecessary, especially when each of these things already has a rich creation API, like the aws SDK as can be seen in it's configuration section https://aws.github.io/aws-sdk-go-v2/docs/configuring-sdk/.

As an integration point, I think it would be better to provide a set of functional options and strip back the abstraction to specific create methods for each catalogue as it is more typical of a Go API. Go tends to encourage a flatter API with fewer abstractions like these.

That said, these abstractions may make sense in the future, but I again function options are more common in Go.

I have moved `NewGlueCatalog(awscfg aws.Config) *GlueCatalog` to use functional options for the AWS configuration as it is cleaner, more extensible, and removes this coupling directly to aws.Config.

@zeroshade Just my two cents, and hopefully explains why I kept this catalogue implementation simple.

Reply:
> As an integration point, I think it would be better to provide a set of functional options and strip back the abstraction to specific create methods for each catalogue as it is more typical of a Go API. Go tends to encourage a flatter API with fewer abstractions like these.

That's fair and I'm fine with that. Perhaps in the future we'd add the more generic abstraction to help people convert from existing iceberg libraries but a per-catalog-type `New` functions will work for now.